File: 2000/2170.cpp
#include<iostream>
#include<algorithm>
using namespace std;

struct Line {
	int st, en;
} line[1000001];

int n;

bool cmp(Line a, Line b) {
	if (a.st < b.st) {
		return true;
	}
	if(a.st == b.st && a.en < b.en){
		return true;
	}
	return false;
}

void input() {
	cin >> n;
	for (int i = 0; i < n; ++i) {
		cin >> line[i].st >> line[i].en;
	}
}

void solution() {
	input();
	sort(line, line + n, cmp);
	int start = line[0].st, end =line[0].en;
	int answer = 0;
	for (int i = 1; i < n; ++i) {
		//완전히 겹치는 경우
		if (line[i].en <= end) continue;
		if (end >= line[i].st) {
			end = line[i].en;
		}
		else {
			answer += end - start;
			start = line[i].st;
			end = line[i].en;
		}
	}
	answer += end - start;
	cout << answer;
}

int main() {
	cin.tie(0), cout.tie(0);
	ios_base::sync_with_stdio(0);
	solution();
	return 0;
}

File: 2000/2170.java
import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	static int n, k;
	static int[][] a;
	
    static void input() throws IOException {
    	n = Integer.parseInt(br.readLine());
    	a = new int[n][2];
    	for(int i = 0;i < n; ++i) {
    		st = new StringTokenizer(br.readLine());
    		a[i][0] = Integer.parseInt(st.nextToken());
    		a[i][1] = Integer.parseInt(st.nextToken());
    	}
    }

    static void solution() {
    	Arrays.sort(a, new Comparator<int[]>() {
			@Override
			public int compare(int[] o1, int[] o2) {
				if(o1[0] == o2[0]) return Integer.compare(o1[1], o2[1]);
				return Integer.compare(o1[0], o2[0]);
			}
    	});
    	int answer = 0;
    	int start = a[0][0], end = a[0][1];
    	for(int i = 1;i < n; ++i) {
    		if(a[i][0] <= end) end = Math.max(end, a[i][1]);
    		else {
    			answer += end - start;
    			start = a[i][0];
    			end = a[i][1];
    		}
    	}
    	answer += end - start;
    	System.out.println(answer);
    }
    
    public static void main(String[] args) throws IOException {
   		input();
   		solution();
    }
}
